File: R/ts-query.R
#' @include ts-ast.R pampa.R from-rust.R
NULL

#' Run a tree-sitter `.scm` query against QMD source
#'
#' `r lifecycle::badge("experimental")`
#'
#' Power-user escape hatch alongside [`select_nodes()`]. Compiles
#' `query_text` against the tree-sitter-qmd grammar and returns one
#' entry per match, with captures keyed by name.
#'
#' This bypasses the tidyselect-style mask entirely; use [`select_nodes()`]
#' if a CSS/predicate query is enough. `ts_query()` is the right tool when
#' you need full structural pattern matching, captures, or predicates
#' from the tree-sitter query language. Only the `#eq?`, `#match?`, and
#' `#any-of?` predicate families (and their `#not-`/`#any-` variants) are
#' evaluated by the matcher; any other predicate compiles but is ignored,
#' with a warning.
#'
#' @param input Either a [`ts_tree`] (re-renders to QMD via [`to_qmd()`]
#'   then re-parses), a single string treated as text/file path (per
#'   [`parse_qmd()`]'s rules), or raw text.
#' @param query_text A tree-sitter query string (S-expression form).
#'   See <https://github.com/quarto-dev/q2/tree/main/crates/tree-sitter-qmd/tree-sitter-markdown/queries/>
#'   for live examples.
#' @return A list with one element per match. Each match is itself a
#'   named list whose names are the capture identifiers in `query_text`
#'   and whose values are [`ts_node`] objects. A quantified capture
#'   appears once per captured node, so a match can carry repeated
#'   names; use `m[names(m) == "x"]` to collect them all. Returns
#'   `list()` when no matches are found.
#' @examples
#' \dontrun{
#' ts = parse_qmd("# Heading\n\nbody\n", ast = "ts")
#' ts_query(ts, "(atx_heading) @h")
#' }
#' @export
ts_query = function(input, query_text) {
  stopifnot(is.character(query_text), length(query_text) == 1L, !is.na(query_text))

  text = if (S7::S7_inherits(input, ts_tree)) {
    to_qmd(input)
  } else if (is.character(input) && length(input) == 1L && !is.na(input)) {
    pampa_read_input(input)$text
  } else {
    stop("ts_query(): input must be a ts_tree, a string, or a file path",
         call. = FALSE)
  }

  raw = ts_query_impl(text, query_text)
  if (!is.null(raw$error)) {
    stop("ts_query(): query compilation failed: ", raw$error, call. = FALSE)
  }
  if (length(raw$unsupported)) {
    rlang::warn(paste0(
      "ts_query(): the matcher only evaluates the #eq?, #match?, and ",
      "#any-of? predicate families; ignored: ",
      paste(unlist(raw$unsupported), collapse = ", ")
    ))
  }

  purrr::map(raw$matches %||% list(), function(m) {
    purrr::map(m, ts_node_from_list)
  })
}

File: src/rust/src/ts_query.rs
use crate::ts_ast_to_r;
use extendr_api::prelude::*;
use streaming_iterator::StreamingIterator;
use tree_sitter::{Query, QueryCursor};
use tree_sitter_qmd::{LANGUAGE, MarkdownParser};

/// Compile `query_text` against tree-sitter-qmd and run it on `text`.
/// Returns a tagged list with `matches` (a list of named lists, one per
/// match, keyed by capture name) and `error` (NULL on success, a
/// human-readable string on query compile failure).
pub fn run_ts_query(text: &str, query_text: &str) -> List {
    let language: tree_sitter::Language = LANGUAGE.into();

    let query = match Query::new(&language, query_text) {
        Ok(q) => q,
        Err(e) => {
            return list!(matches = NULL, error = format!("{}", e), unsupported = NULL);
        }
    };

    // The cursor only evaluates the text predicates the Rust binding
    // implements (#eq?, #match?, #any-of? and their variants). Anything else
    // compiles as a "general" predicate and is silently ignored by the
    // matcher, so surface the names and let the R side warn.
    let mut unsupported: Vec<String> = Vec::new();
    for i in 0..query.pattern_count() {
        for p in query.general_predicates(i) {
            let name = format!("#{}", p.operator);
            if !unsupported.contains(&name) {
                unsupported.push(name);
            }
        }
    }

    // `MarkdownParser::parse` sets the block grammar itself; `language` above
    // is kept only for `Query::new`.
    let mut parser = MarkdownParser::default();

    let padded = ts_ast_to_r::ensure_trailing_newline(text.as_bytes());
    let bytes: &[u8] = padded.as_ref();

    let tree = match parser.parse(bytes, None) {
        Some(t) => t,
        None => {
            return list!(matches = List::new(0), error = NULL, unsupported = NULL);
        }
    };

    let capture_names: Vec<&str> = query.capture_names().iter().copied().collect();

    let top = tree.walk_cursor();
    let root = top.node();

    let mut cursor = QueryCursor::new();
    let mut matches_iter = cursor.matches(&query, root, bytes);

    let mut out_matches: Vec<Robj> = Vec::new();
    while let Some(m) = matches_iter.next() {
        let mut names: Vec<&str> = Vec::with_capacity(m.captures.len());
        let mut values: Vec<Robj> = Vec::with_capacity(m.captures.len());
        for cap in m.captures.iter() {
            let name = capture_names.get(cap.index as usize).copied().unwrap_or("");
            let mut walker = cap.node.walk();
            let node_value = ts_ast_to_r::node_to_r_at(&mut walker, bytes);
            names.push(name);
            values.push(node_value);
        }
        let mut entry = List::from_values(values);
        let _ = entry.set_names(names);
        out_matches.push(entry.into());
    }

    let unsupported_r: Robj = if unsupported.is_empty() {
        ().into()
    } else {
        unsupported.into()
    };
    list!(
        matches = List::from_values(out_matches),
        error = NULL,
        unsupported = unsupported_r
    )
}
